Implement fetching consumer_secret from Salesforce connection

kanagaraj-dhanapal-89 opened the pull request:
This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

kanagaraj-dhanapal-89 committed:
Salesforce: Implement fetching consumer_secret from Airflow Connections

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

kanagaraj-dhanapal-89 committed:
Update salesforce.py

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

eladkal commented on providers/src/airflow/providers/salesforce/hooks/salesforce.py:
```diff
@@ -151,6 +151,7 @@ def conn(self) -> api.Salesforce:
             session=self.session,
             client_id=self._get_field(extras, "client_id") or None,
             consumer_key=self._get_field(extras, "consumer_key") or None,
+            consumer_secret=self._get_field(extras, "consumer_secret") or None,
```
Where is the code part that use it?

kanagaraj-dhanapal-89 replied:
We have a custom Salesforce API Fetch Operator that fetches Salesforce objects using SQL queries. To use this operator, we need to pass the consumer key from the Airflow Connections extra

`class SalesforceApiDataExtractionOperator(BaseOperator):

    def __init__(self,
                 task_id: str,
                 query: str,
                 upload_path:str,
                 gcp_conn_id: str,
                 salesforce_conn_id: str = "salesforce_conn_id",
                 include_deleted: bool = False,
                 query_params: dict | None = None,
                 **kwargs

    ):
        self.upload_path = upload_path
        self.gcp_conn_id = gcp_conn_id
        self.task_id = task_id
        self.salesforce_conn_id = salesforce_conn_id
        self.include_deleted = include_deleted
        self.query_params = query_params
        self.query = query

        super().__init__(task_id=task_id,**kwargs)

    def __get_local_storage_location(self):
        _base_loc = os.path.join(os.getenv("HOME"), self.task_id)
        if not os.path.exists(_base_loc):
            os.mkdir(_base_loc)
        return _base_loc

    def __clean_up(self, df):
        self.log.info("started data clean up")
        df = df.replace("~", " ").replace("\n", " ").replace("\r", "")
        self.log.info("data clean up completed")
        return df
    def __write_local_file(self, df:DataFrame):

        file_name = os.path.basename(self.upload_path)
        local_folder_path = self.__get_local_storage_location().rstrip('/')
        local_file_path = f"{local_folder_path}/{file_name}"
        df.to_csv(local_file_path, header=False, sep="~")

        return local_file_path

    def __move_to_gcs(self, file_path):
        hook = GCSHook(gcp_conn_id=self.gcp_conn_id)
        bucket_name = self.upload_path.replace("gs://", "").split("/")[0]
        self.log.info(f"bucket: {bucket_name}")
        gcs_file_path = "/".join(self.upload_path.replace("gs://", "").split("/")[1:])
        self.log.info(f"upload file: {gcs_file_path}")
        hook.upload(
            bucket_name=bucket_name,
            object_name=gcs_file_path,
            filename=file_path
        )

    def execute(self, context: Context):
        if context["params"].get("sf_filter", None):
            query = self.query + " " + context["params"]["sf_filter"]
        else:
            query = self.query
        sf_hook = SalesforceHook(salesforce_conn_id=self.salesforce_conn_id)
        results = sf_hook.make_query(
            query=query,
            include_deleted=self.include_deleted,
            query_params=self.query_params
        )
        df = sf_hook.object_to_df(results["records"])
        df = self.__clean_up(df)
        file_path = self.__write_local_file(df)
        self.__move_to_gcs(file_path)`

eladkal replied:
You'll need also to update docs 
https://github.com/apache/airflow/blob/5c0e98cc770b4f055dbd1c0b60ccbd69f3166da7/docs/apache-airflow-providers-salesforce/connections/salesforce.rst#L46

and `get_connection_form_widgets` function.

kanagaraj-dhanapal-89 replied:
Code implemented and documentation updated.

eladkal replied:
ah we need also test coverage.
Probably easiest is jus to update `test_get_conn_jwt_auth`

https://github.com/apache/airflow/blob/857ca4c06c9008593674cabdd28d3c30e3e7f97b/providers/tests/salesforce/hooks/test_salesforce.py#L155

kanagaraj-dhanapal-89 replied:
updated the test case

eladkal reviewed (changes requested):
Please ammend the commit and PR title with meeaningful description

Commit:
Merge remote-tracking branch 'origin/kanagaraj-dhanapal-89-patch-1' into kanagaraj-dhanapal-89-patch-1

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections and connection form widgets

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections and connection form widgets

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections and connection form widgets

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections and connection form widgets + UTC's update

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections and connection form widgets

This PR introduces a new feature that allows users to retrieve the consumer_secret directly from the Connections service.

Commit:
Merge remote-tracking branch 'origin/kanagaraj-dhanapal-89-patch-1' into kanagaraj-dhanapal-89-patch-1

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

kanagaraj-dhanapal-89 committed:
Merge branch 'main' into kanagaraj-dhanapal-89-patch-1

Commit:
Salesforce: Implement fetching consumer_secret from Airflow Connections test cases changes
